torchdata/datapipes/iter/transform/callable.py

```diff
@@ -283,3 +284,84 @@ def __len__(self) -> int:
+class FlattenIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    returns a flattened copy of the input DataPipe based on provided indices (functional name: ``flatten``).
+
+    Args:
+        datapipe: IterDataPipe with iterable elements
+        indices: a single index/key for the item to flatten from an iterator item or a list of indices/keys to be flattened
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper([(0, 10, (100, 1000)), (1, 11, (111, 1001)), (2, 12, (122, 1002)), (3, 13, (133, 1003)), (4, 14, (144, 1004))])
+        >>> flatten_dp = dp.flatten(2)
+        >>> list(flatten_dp)
+        [(0, 10, 100, 1000), (1, 11, 111, 1001), (2, 12, 122, 1002), (3, 13, 133, 1003), (4, 14, 144, 1004)]
+    """
+    datapipe: IterDataPipe
+    indices = None
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe,
+        indices: Optional[Union[Hashable, List[Hashable]]] = None,
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+        if indices:
+            if isinstance(indices, list):
+                self.indices = set(indices)
+            else:
+                self.indices = {indices}
+
+    def __iter__(self) -> Iterator[T_co]:
+        for old_item in self.datapipe:
+            if self.indices:
+                if isinstance(old_item, dict):
+                    new_item = {}  # type: ignore[assignment]
+                    for k, v in old_item.items():
+                        if k in self.indices:
+                            for k_sub, v_sub in v.items():
+                                new_item[k_sub] = v_sub
+                        else:
+                            new_item[k] = v
+                else:
+                    is_tuple = False
+                    new_item = []  # type: ignore[assignment]
+                    if isinstance(old_item, tuple):
+                        is_tuple = True
+                        old_item = list(old_item)
+                    for i, item in enumerate(old_item):
+                        if i in self.indices:
+                            new_item.extend(list(item))  # type: ignore[attr-defined]
+                        else:
+                            new_item.append(item)  # type: ignore[attr-defined]
+                    if is_tuple:
+                        new_item = tuple(new_item)  # type: ignore[assignment]
+
+                # check to make sure all indices requested were in the item. warn if not
+                try:
+                    for index in self.indices:
+                        old_item[index]
+                except (IndexError, KeyError):
+                    warnings.warn(
+                        "At least one index in the filter is not present in the item being returned,"
+                        " please be aware that expected columns/keys may be missing."
+                    )
+
+                yield new_item  # type: ignore[misc]
+
+            else:
```

Comment:
@VitalyFedyunin does this look like the default behavior you wanted? One level of automatic flattening but more like an iterator created across that level, so you get one item at a time from that flattened list? It's what was described by @NicolasHug here: https://github.com/pytorch/data/issues/648 (and probably why he wanted to name it chainer).

Reply:
The proposed solution looks like `unbatch` to me. 

For default behaviour I would rather do something like:

```python
list(dp) # [ (1, (2,3), 4), (5, (6,7), 8) ]
dp = dp.flatten()
list(dp) # [ (1, 2, 3,4), (5,6,7,8) ]
```

Meaning expand all nested fields by checking something like:

```
from collections.abc import Iterable

if not isinstance(x, Iterable):
  x = [x]
```

Reply:
We probably want to exclude `str` class. Do avoid expanding strings to characters.

Reply:
I can make that change. It won't solve the problem described in Nicolas's request though at least how I've read what he was looking for.

Reply:
It will not. He needs `unbatch` or no-op `flatmap`.

Reply:
@VitalyFedyunin What makes you prefer that default behavior instead of what is implemented?

I am used to this being the default:
```python
list(dp)  # [(1, (2, 3), 4), (5, (6, 7), 8)]
dp = dp.flatten()
list(dp)  # [1, (2, 3), 4, 5, (6, 7), 8]
```
Here is the [`numpy/scipy` implementation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flatten.html) for reference.

I agree that we can have additional arguments `indices` to unpack specific indices rather than the whole element. Just curious about the choice of default here.

Reply:
Updated default per Vitaly's feedback. Lmk if you two decide we should change this. The newest commit shows the new default behavior so you can compare to the previous option.